kythe/rust/fuchsia_extractor/src/bin/main.rs

```diff
@@ -0,0 +1,862 @@
+fn process_files(files: &[PathBuf], options: &Options) -> Result<Vec<PathBuf>> {
+    let mut kzips: Vec<PathBuf> = vec![];
+    for file_name in files {
+        let file = fs::File::open(&file_name)
+            .with_context(|| format!("process_files: while opening file: {:?}", &file_name))?;
+        let analysis: rls_data::Analysis = serde_json::from_reader(file).with_context(|| {
+            format!(
+                "process_files: while parsing save-analysis JSON from file: {:?}",
+                &file_name
+            )
+        })?;
+        let kzip = process_file(&file_name, &options.output_dir, &analysis, options)
+            .with_context(|| format!("process_files: while processing file: {:?}", &file_name))?;
+        kzips.push(kzip);
+    }
+    Ok(kzips)
+}
+
+/// Makes an absolute `path` relative to its `parent`.  For example,
+fn make_relative_to(path: &impl AsRef<Path>, parent: &impl AsRef<Path>) -> Result<PathBuf> {
+    let mut src_components = path.as_ref().components().peekable();
+    let mut dest_components = parent.as_ref().components().peekable();
+
+    // Start from the canonical paths of source and destination.  Remove the maximal shared prefix.
+    // Then assemble the final path from (1) the remaining path components of dest, replaced by
+    // "..", and then remaining path components of source.
+    loop {
+        let src = src_components.peek();
+        let dest = dest_components.peek();
+        match (src, dest) {
+            (Some(s), Some(d)) => {
+                if s != d {
+                    break;
+                }
+                src_components.next();
+                dest_components.next();
+            }
+            _ => {
+                break;
+            }
+        }
+    }
+    let mut result = PathBuf::new();
+    let _ = dest_components.map(|_| result.push("..")).collect::<()>();
+    let _ = src_components.map(|c| result.push(&c)).collect::<()>();
+    Ok(result)
+}
+
+/// The shared general settings for the compilation extractor.
+struct Options {
+    /// The name of this corpus, e.g. "fuchsia".
+    corpus_name: String,
+    /// The programming language this data is for, e.g. "rust"
+    language_name: String,
+    /// The directory in which compilation ran.
+    base_dir: PathBuf,
+    /// The directory to which to write the kzip file outputs.
+    output_dir: PathBuf,
+    /// The comma-separated revisions string, used to fill out the indexed compilation
+    /// protocol buffer field called `index`.
+    revisions: Vec<String>,
+}
+
+fn main() -> Result<()> {
+    // We'd rather use a crate that parses the program arguments directly into the `Options`
+    // struct.  However, for some weird reason our dependencies can not be so aligned to get any
+    // such crate to compile, even including a newer version of `clap`, which supports this kind of
+    // parsing.  So we parse like this.
```

Comment:
Which crate were you trying to use? I can try and get it to work with cargo-raze. A lot of times it just involves setting `gen_buildrs` in Cargo.toml for the dependency

Reply:
I tried to use `structopt`, but the issue I had was that in pulls in `syn`; `syn` would not compile because it is trying to use `proc-macro`, with a combination of features that does not work.  I gave up on figuring that dependency hell out, and just used `clap` as it was already used elsewhere in the Rust code here.

Reply:
I'll resolve this since any improvements we may want to make here can be made separately.